### ISSUE TITLE
Iterator version of AgentExecutor

### ISSUE THREAD

**Issue by SlapDrone** [langchain-ai/langchain#6925]
### Feature request

A colleague and I would like to implement an iterator version of the AgentExecutor. That is, we'd like to expose each intermediate step of the plan/action/observation loop. This should look something like:

```python3
inputs = ... # complex question that requires tool usage, multi-step

for step in agent_executor.iter(inputs=inputs):
    do_stuff_with(step)
```

### Motivation

This would be useful for a few applications. By hooking into agent steps we could:

- Route different thoughts/actions/observations conditionally to:
    - different frontend components 
    - different postprocessing/analysis logic
    - other agents

We could also "pause" agent execution and intervene.

### Your contribution

Yes, we have a PR ~ready.

**Comment by UmerHA** [comment]
Nice! fyi: This touches on https://github.com/hwchase17/langchain/issues/5086 and https://github.com/hwchase17/langchain/issues/5448

**Comment by SlapDrone** [comment]
Closed by #6929
